nav2_route/src/breadth_first_search.cpp

```diff
@@ -0,0 +1,176 @@
+{
+  auto iter = graph_.find(index);
+  if (iter != graph_.end()) {
+    return &(iter->second);
+  }
+
+  return &(graph_.emplace(index, Node{index, false}).first->second);
+}
+
+void BreadthFirstSearch::setStart(unsigned int mx, unsigned int my)
+{
+  start_ = addToGraph(costmap_->getIndex(mx, my));
+}
+
+void BreadthFirstSearch::setGoals(std::vector<unsigned int> mxs, std::vector<unsigned int> mys)
+{
+  goals_.clear();
+  for (unsigned int i = 0; i < (mxs.size() && mys.size()); ++i) {
+    goals_.emplace_back(addToGraph(costmap_->getIndex(mxs[i], mys[i])));
+  }
+}
+
+bool BreadthFirstSearch::search(Coordinates & closest_goal)
+{
+  // clear the graph
+  std::unordered_map<unsigned int, Node> empty_graph;
+  std::swap(graph_, empty_graph);
+
+  std::queue<NodePtr> queue;
+
+  start_->visited = true;
+  queue.push(start_);
+
+  std::cout << "Start index " << start_->index << std::endl;
+  std::cout << "Goal index " << goals_[0]->index << std::endl;
+
+  int iter = 0;
+  while (!queue.empty()) {
+    iter += 1;
+    std::cout << "Iteration " << iter << std::endl;
+    std::cout << "Queue Size: " << queue.size() << std::endl;
+
+    auto & current = queue.front();
+
+    std::cout << "grab front" << std::endl;
+    std::cout << std::endl;
+
+    Coordinates current_coord = getCoords(current->index);
+    std::cout << "CURRENT NODE: " << current_coord.x << " " << current_coord.y << std::endl;
+
+    queue.pop();
+
+    // Check goals
+    for (const auto & goal : goals_) {
+      std::cout << "Goal index " << goal->index << std::endl;
+      std::cout << "current index " << current->index << std::endl;
+      if (current->index == goal->index) {
+        std::cout << "Found goal" << std::endl;
+        closest_goal = getCoords(current->index);
+        return true;
+      }
+    }
+    std::cout << "Checked goals" << std::endl;
+
+    NodeVector neighbors;
+    getNeighbors(current->index, neighbors);
+    std::cout << "Got neigbors" << std::endl;
+
+    for (const auto neighbor : neighbors) {
+      if (!neighbor->visited) {
+        neighbor->visited = true;
+        queue.push(neighbor);
+
+        std::cout << "Added node: " << neighbor->index << std::endl;
+      }
+    }
+  }
+  return false;
+}
+
+bool BreadthFirstSearch::getNeighbors(unsigned int current, NodeVector & neighbors)
+{
+  const Coordinates p_coord = getCoords(current);
+
+  unsigned int index;
+  Coordinates c_coord{0, 0};
+
+  std::cout << "grid offset size " << neighbors_grid_offsets.size() << std::endl;
+  for (const int & neighbors_grid_offset : neighbors_grid_offsets) {
+    std::cout << "Getting neighbors" << std::endl;
+    index = current + neighbors_grid_offset;
+
+    c_coord = getCoords(index);
+
+    // Check for wrap around conditions
+    if (std::fabs(p_coord.x - c_coord.x) > 1 || std::fabs(p_coord.y - c_coord.y) > 1) {
+      std::cout << "Index wraps" << std::endl;
+      continue;
+    }
+
+    // Check for out of bounds
+    if (index >= max_index_) {
```

Comment:
Or < 0

Reply:
it is a unsigned int. If we are below zero C++ has failed us :)